QHyper/problems/community_detection.py

```diff
@@ -0,0 +1,193 @@
+from dataclasses import dataclass, field
+from typing import Any, Iterable, Tuple, cast
+
+import networkx as nx
+import numpy as np
+import sympy
+from QHyper.hyperparameter_gen.parser import Expression
+from QHyper.problems.base import Problem
+from QHyper.util import VARIABLES
+from sympy.core.expr import Expr
+
+
+@dataclass
+class Network:
+    graph: nx.Graph
+    resolution: float = field(default=1)
+    modularity_matrix: np.ndarray = field(init=False)
+    weight: str | None = field(default=None)
+
+    def __post_init__(self) -> None:
+        self.modularity_matrix = self.calculate_modularity_matrix()
+
+    def calculate_modularity_matrix(self) -> np.ndarray:
+        adj_matrix: np.ndarray = nx.to_numpy_array(
+            self.graph, weight=self.weight
+        )
+        degree_matrix: np.ndarray = adj_matrix.sum(axis=1)
+        m: int = np.sum(degree_matrix)
+        return (
+            adj_matrix
+            - self.resolution * np.outer(degree_matrix, degree_matrix) / m
+        )
+
+
+class KarateClubNetwork(Network):
+    def __init__(self, resolution: float = 1):
+        super().__init__(nx.karate_club_graph(), resolution=resolution)
+
+
+class BrainNetwork(Network):
+    def __init__(
+        self,
+        input_data_dir: str,
+        input_data_name: str,
+        delimiter: str = "	",
+        resolution: int = 1,
+    ):
+        adj_matrix = np.genfromtxt(
+            f"{input_data_dir}/{input_data_name}.csv", delimiter=delimiter
+        )
+        super().__init__(
+            nx.from_numpy_matrix(adj_matrix), resolution=resolution
+        )
+
+
+class CommunityDetectionProblem(Problem):
+    """
+    Problem class instance
+    - objective function for network community detection
+
+    Attributes:
+    ----------
+    cases: int
+        number of communities into which the graph shall be divided
+        (default 2)
+    G: networkx graph
+        networkx implementation of graph
+    B: networkx modularity matrix
+        networkx implementation of modularity matrix
+
+    objective_function : Expression
+        objective function in SymPy syntax
+    constraints : list[Expression]
+        list of problem constraints in SymPy syntax, (default [])
+    variables : int
+        number of qubits in the circuit, equal to number of nodes
+        in the graph
+    """
+
+    def __init__(self, network_data: Network, N_communities: int = 2) -> None:
```

Comment:
I don't like starting variable name with capital letter, and imo this N doesn't add anything
```suggestion
    def __init__(self, network_data: Network, communities: int = 2) -> None:
```